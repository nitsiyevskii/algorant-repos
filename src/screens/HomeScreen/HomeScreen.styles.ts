import { createStyles } from "src/utils/createStyles";

export const useStyles = createStyles((theme) => ({
  safeArea: {
    flex: 1,
    backgroundColor: theme.colors.background.white,
  },
  container: {
    flex: 1,
    backgroundColor: theme.colors.background.primary,
  },
  searchContainer: {
    paddingHorizontal: 16,
    paddingTop: 16,
    paddingBottom: 12,
    backgroundColor: theme.colors.background.white,
    borderBottomWidth: 1,
    borderBottomColor: theme.colors.separator.secondary,
  },
  searchRow: {
    flexDirection: "row",
    gap: 12,
  },
  searchInput: {
    flex: 1,
  },
  listContainer: {
    padding: 16,
  },
}));
