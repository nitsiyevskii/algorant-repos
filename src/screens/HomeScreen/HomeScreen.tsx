import React from "react";
import { View, FlatList } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import { router } from "expo-router";
import { SearchInput } from "src/components/SearchInput/SearchInput";
import { LoadingSpinner } from "src/components/LoadingSpinner/LoadingSpinner";
import { EmptyState } from "src/components/EmptyState/EmptyState";
import { FilterButton } from "./components/FilterButton/FilterButton";
import { RepositoryCard } from "./components/RepositoryCard/RepositoryCard";
import { GitHubRepository } from "src/types";
import { useRepositories } from "./hooks/useRepositories";
import { useRepositorySearch } from "./hooks/useRepositorySearch";
import { useFavorites } from "./hooks/useFavorites";
import { useStyles } from "./HomeScreen.styles";

const HomeScreen = () => {
  const styles = useStyles();
  
  // Custom hooks for business logic
  const { repositories, isLoading, error } = useRepositories();
  const { searchQuery, setSearchQuery, filteredRepositories } = useRepositorySearch(repositories);
  const { isFavorite, toggleFavorite } = useFavorites();

  const handleFilterPress = () => {
    router.push('/filter');
  };

  const renderRepository = ({ item }: { item: GitHubRepository }) => (
    <RepositoryCard
      repository={item}
      isFavorite={isFavorite(item.id)}
      onToggleFavorite={() => toggleFavorite(item.id)}
    />
  );

  const renderEmpty = () => (
    <EmptyState
      icon="search"
      message={searchQuery ? "No repositories found" : "No repositories available"}
      description={searchQuery ? "Try adjusting your search" : undefined}
    />
  );

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (error) {
    return (
      <EmptyState
        icon="alert-circle"
        message="Failed to load repositories"
        description={error}
      />
    );
  }

  return (
    <SafeAreaView style={styles.safeArea} edges={['top']}>
      <View style={styles.container}>
        <View style={styles.searchContainer}>
          <View style={styles.searchRow}>
            <View style={styles.searchInput}>
              <SearchInput
                value={searchQuery}
                onChangeText={setSearchQuery}
                placeholder="Search repositories..."
              />
            </View>
            <FilterButton onPress={handleFilterPress} />
          </View>
        </View>

        <FlatList
          data={filteredRepositories}
          renderItem={renderRepository}
          keyExtractor={(item) => item.id.toString()}
          contentContainerStyle={styles.listContainer}
          ListEmptyComponent={renderEmpty}
        />
      </View>
    </SafeAreaView>
  );
};

export default HomeScreen;
