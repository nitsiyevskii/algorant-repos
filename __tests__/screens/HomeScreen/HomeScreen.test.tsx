import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import { useDispatch, useSelector } from 'react-redux';
import HomeScreen from 'src/screens/HomeScreen/HomeScreen';

jest.mock('react-redux');

const mockDispatch = jest.fn();
const mockUseDispatch = useDispatch as unknown as jest.Mock;
const mockUseSelector = useSelector as unknown as jest.Mock;

const mockRepositories = [
  {
    id: 1,
    name: 'test-repo',
    full_name: 'algorand/test-repo',
    owner: {
      login: 'algorand',
      avatar_url: 'https://test.com/avatar.png',
      html_url: 'https://github.com/algorand',
    },
    html_url: 'https://github.com/algorand/test-repo',
    description: 'Test repository',
    fork: false,
    created_at: '2023-01-01T00:00:00Z',
    updated_at: '2023-01-02T00:00:00Z',
    pushed_at: '2023-01-02T00:00:00Z',
    stargazers_count: 100,
    watchers_count: 50,
    language: 'TypeScript',
    forks_count: 10,
    open_issues_count: 5,
    topics: ['blockchain', 'algorand'],
    visibility: 'public',
    default_branch: 'main',
  },
];

describe('HomeScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseDispatch.mockReturnValue(mockDispatch);
    
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: mockRepositories,
          favorites: [],
          isLoading: false,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });
  });

  it('renders correctly', () => {
    const { getByPlaceholderText } = render(<HomeScreen />);
    expect(getByPlaceholderText('Search repositories...')).toBeTruthy();
  });

  it('shows loading spinner when loading', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: true,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { getByTestId } = render(<HomeScreen />);
    expect(getByTestId('loading-spinner')).toBeTruthy();
  });

  it('shows error state when error occurs', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: false,
          error: 'Failed to fetch',
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { getByText } = render(<HomeScreen />);
    expect(getByText('Failed to load repositories')).toBeTruthy();
    expect(getByText('Failed to fetch')).toBeTruthy();
  });

  it('shows empty state when no repositories', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: false,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { getByText } = render(<HomeScreen />);
    expect(getByText('No repositories available')).toBeTruthy();
  });

  it('renders repository list', async () => {
    const { getByText } = render(<HomeScreen />);
    
    await waitFor(() => {
      expect(getByText('test-repo')).toBeTruthy();
      expect(getByText('algorand')).toBeTruthy();
    });
  });

  it('renders search input', () => {
    const { getByPlaceholderText } = render(<HomeScreen />);
    expect(getByPlaceholderText('Search repositories...')).toBeTruthy();
  });

  it('renders filter button', () => {
    const { getByText } = render(<HomeScreen />);
    expect(getByText('Filter')).toBeTruthy();
  });

  it('dispatches fetchRepositories on mount', () => {
    render(<HomeScreen />);
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<HomeScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when loading', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: true,
          error: null,
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { toJSON } = render(<HomeScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when error', () => {
    mockUseSelector.mockImplementation((selector) => {
      const state = {
        repositories: {
          items: [],
          favorites: [],
          isLoading: false,
          error: 'Test error',
        },
        filters: {
          organizations: [],
          languages: [],
          stars: { min: null, max: null },
          forks: { min: null, max: null },
          searchQuery: '',
        },
      };
      return selector(state);
    });

    const { toJSON } = render(<HomeScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
