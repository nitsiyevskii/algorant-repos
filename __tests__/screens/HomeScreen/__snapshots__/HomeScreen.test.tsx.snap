// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen matches snapshot 1`] = `
<View
  edges={
    [
      "top",
    ]
  }
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#F2F2F7",
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#E5E5EA",
          "borderBottomWidth": 1,
          "paddingBottom": 12,
          "paddingHorizontal": 16,
          "paddingTop": 16,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "gap": 12,
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#F2F2F7",
                  "borderColor": "#C7C7CC",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 48,
                  "paddingHorizontal": 12,
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <View
              style={
                {
                  "marginRight": 8,
                }
              }
            >
              <Text
                color="#8E8E93"
                name="search"
                size={20}
                testID="icon"
              >
                search
              </Text>
            </View>
            <TextInput
              autoCapitalize="none"
              autoCorrect={false}
              editable={true}
              onChangeText={[Function]}
              placeholder="Search repositories..."
              placeholderTextColor="#8E8E93"
              style={
                {
                  "color": "#000000",
                  "flex": 1,
                  "fontSize": 16,
                  "height": "100%",
                  "paddingVertical": 0,
                }
              }
              testID="text-input"
              value=""
            />
          </View>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#F2F2F7",
              "borderColor": "#C7C7CC",
              "borderRadius": 10,
              "borderWidth": 1,
              "flexDirection": "row",
              "gap": 8,
              "height": 48,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 16,
            }
          }
        >
          <View>
            <Text
              color="#000000"
              name="filter"
              size={20}
              testID="icon"
            >
              filter
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "color": "#000000",
                  "fontSize": 16,
                  "fontWeight": "500",
                },
                false,
              ]
            }
          >
            Filter
          </Text>
        </View>
      </View>
    </View>
    <RCTScrollView
      ListEmptyComponent={[Function]}
      contentContainerStyle={
        {
          "padding": 16,
        }
      }
      data={
        [
          {
            "created_at": "2023-01-01T00:00:00Z",
            "default_branch": "main",
            "description": "Test repository",
            "fork": false,
            "forks_count": 10,
            "full_name": "algorand/test-repo",
            "html_url": "https://github.com/algorand/test-repo",
            "id": 1,
            "language": "TypeScript",
            "name": "test-repo",
            "open_issues_count": 5,
            "owner": {
              "avatar_url": "https://test.com/avatar.png",
              "html_url": "https://github.com/algorand",
              "login": "algorand",
            },
            "pushed_at": "2023-01-02T00:00:00Z",
            "stargazers_count": 100,
            "topics": [
              "blockchain",
              "algorand",
            ],
            "updated_at": "2023-01-02T00:00:00Z",
            "visibility": "public",
            "watchers_count": 50,
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#FFFFFF",
                "borderColor": "#C6C6C8",
                "borderRadius": 12,
                "borderWidth": 1,
                "marginBottom": 12,
                "opacity": 1,
                "padding": 16,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginBottom": 8,
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "marginRight": 12,
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#000000",
                      "fontSize": 18,
                      "fontWeight": "600",
                      "marginBottom": 4,
                    }
                  }
                >
                  test-repo
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#3C3C43",
                      "fontSize": 14,
                    }
                  }
                >
                  algorand
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": false,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#F2F2F7",
                    "borderRadius": 20,
                    "height": 40,
                    "justifyContent": "center",
                    "opacity": 1,
                    "width": 40,
                  }
                }
                testID="favorite-button"
              >
                <Text
                  color="#000000"
                  name="heart-outline"
                  size={24}
                  testID="icon"
                >
                  heart-outline
                </Text>
              </View>
            </View>
            <Text
              numberOfLines={2}
              style={
                {
                  "color": "#3C3C43",
                  "fontSize": 16,
                  "lineHeight": 20,
                  "marginBottom": 12,
                }
              }
            >
              Test repository
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 16,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <Text
                  color="#8E8E93"
                  name="star"
                  size={16}
                  testID="icon"
                >
                  star
                </Text>
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  100
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <View
                  style={
                    [
                      {
                        "borderRadius": 6,
                        "height": 12,
                        "width": 12,
                      },
                      {
                        "backgroundColor": "#007AFF",
                      },
                    ]
                  }
                />
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  TypeScript
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                  }
                }
              >
                <Text
                  color="#8E8E93"
                  name="git-branch"
                  size={16}
                  testID="icon"
                >
                  git-branch
                </Text>
                <Text
                  style={
                    {
                      "color": "#8E8E93",
                      "fontSize": 14,
                    }
                  }
                >
                  10
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;

exports[`HomeScreen matches snapshot when error 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 32,
    }
  }
>
  <Text
    color="#8E8E93"
    name="alert-circle"
    size={64}
    testID="icon"
  >
    alert-circle
  </Text>
  <Text
    style={
      {
        "color": "#3C3C43",
        "fontSize": 18,
        "fontWeight": "500",
        "marginTop": 16,
        "textAlign": "center",
      }
    }
  >
    Failed to load repositories
  </Text>
  <Text
    style={
      {
        "color": "#8E8E93",
        "fontSize": 16,
        "marginTop": 8,
        "textAlign": "center",
      }
    }
  >
    Test error
  </Text>
</View>
`;

exports[`HomeScreen matches snapshot when loading 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F2F2F7",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loading-spinner"
>
  <ActivityIndicator
    color="#007AFF"
    size="large"
    testID="activity-indicator"
  />
</View>
`;
